### ISSUE TITLE
provisioner/shell: properly retry cleanup command to work through reboots

### ISSUE THREAD

**Issue by DanHam** [mitchellh/packer#4137]
Issuing a reboot command within a provisioning script causes the Packer build to fail. The issue is present in versions 0.11.0 and 0.11.1.dev and reproducable with both Virtualbox and VMware builders. The same template and script combination works perfectly with version 0.10.2. 

The issue appears to be caused by Packer attempting to remove the provisioning script after the machine has begun to reboot.

Debug log output: 
- From VMware builder: https://gist.github.com/DanHam/1866b6bd6ef3daca26b238e452c9c174

Template and scripts required to replicate the issue: 
- See: https://github.com/DanHam/packer-issue.git

Packer version: 0.11.0 and 0.11.1.dev
Host platform: Mac OS X 10.11.6
Virtualbox: 5.0.26
VMware Fusion: Professional Version 8.5.1 (4543325)

*Many thanks!*

**Comment by rickard-von-essen** [comment]
As a workaround you can set `"skip_clean": false`, see [shell - skip_clean](https://www.packer.io/docs/provisioners/shell.html#skip_clean).

**Comment by jinesh-choksi** [comment]
In case it helps you, here is how we apply updates and reboot in CentOS/RHEL:

```
#!/usr/bin/env bash

echo "##############################################################################"
echo "# 03_install_system_updates_and_reboot.sh                                    #"
echo "##############################################################################"

echo "Clearing YUM meta data."
yum clean all

echo "Updating YUM meta data."
yum makecache

echo "Updating all packages installed on the system."
yum -y update

# Shutdown the SSHD service before the reboot is initiated to prevent Packer
# from trying to SSH in and executing the outstanding shell provisioning
# scripts.
# Source: https://github.com/mitchellh/packer/issues/3487#issue-152469296
echo "Shutting down the SSHD service and rebooting:"
systemctl stop sshd.service

echo "Rebooting:"
reboot

# The sleep statement is required to block Packer from executing the next
# script. Sometimes, when executing a command like reboot, the shell script
# will return and Packer will start executing the next one before SSH actually
# quits and the machine restarts. For this, put a long sleep after the reboot
# so that SSH will eventually be killed automatically.
sleep 300s
```

With the above, you also won't need to use the "pause_before" stanza in the Packer template.

**Comment by rickard-von-essen** [comment]
The simple solution is that you just add [`skip_clean`](https://www.packer.io/docs/provisioners/shell.html#skip_clean) to the provisioner which reboots. And add [`pause_before`](https://www.packer.io/docs/templates/provisioners.html) to the script that needs to start after the shutdown (it will wait until ssh is available).

@mwhooker I don't think this is a bug.

**Comment by jinesh-choksi** [comment]
@rickard-von-essen Yes that will fix his current issue but depending on his requirements he may face another issue right after. e.g. if he is required to keep the `pause_before` time to be as small as possible since he needs to reduce the overall build time of the image, then he will find that Packer has some issues with handling reboots. Please have a look at the discussion in issue #3487 to get a clearer picture.

I agree that a large enough `pause_before` value is definitely a possible fix if you are comfortable with the delay it adds to the builds.

**Comment by rickard-von-essen** [comment]
> keep the pause_before time to be as small as possible since he needs to reduce the overall build time of the image

I disagree setting `pause_before` won't add any build time when used after a reboot. The following sequence should work:

```json
{
  "type": "shell",
  "inline": [ "sudo reboot" ],
  "skip_clean": true
},
{
  "type": "shell",
  "pause_before": "15s",
  "inline": [ "uptime" ]
}
```

Since rebooting take longer time than 15s this doesn't add any build time.

@mwhooker Except for clarifying the docs a bit I'm sceptical about these issues. I'm voting for close.

**Comment by DanHam** [comment]
@rickard-von-essen Thanks for the hint regarding the "skip_clean" workaround. I will give this a try over the weekend and see if it allows my build to complete successfully in the meantime.

> Except for clarifying the docs a bit I'm sceptical about these issues. I'm voting for close.

Can you clarify what you are sceptical about? I would like to raise some objections with the vote to close and the suggestion that this isn't a bug: 

- Packer 0.10.2 handled these reboots perfectly** without the need for "skip_clean" in the template or by issuing commands to shutdown sshd from within the script. This suggests the behaviour we are now seeing has been introduced by changes made somewhere between 0.10.2 and 0.11.0.

- While "skip_clean" may prove to be a valid workaround, it means we are left with Packer build scripts in the final image which is _**very**_ undesirable. Preserving the default behaviour of having Packer clean up after itself is far preferable.

- The example template and scripts I created to reproduce the issue were (clearly!) somewhat of a simplification. The actual update script I use makes a decision as to whether or not to reboot the machine based on the updates it has installed (it does this by grepping for updates known to require a reboot e.g. kernel or glibc updates). As such, if a reboot command isn't actually issued the inclusion of "skip_clean" isn't actually necessary. However, the script would have no way of telling Packer this is the case. The actual script I use to update my CentOS boxes can be seen here: https://gist.github.com/DanHam/4dfd58e56993e938efa183a9471ff62e. 

- In the grand scheme of things I'm not overly concerned with the potential to add a bit of additional build time due to a "pause_before" in the next provisioner. I've actually found a "pause_before" of _just 5 seconds_ works perfectly well and, as you rightly point out, if the machine _is_ actually rebooted by the previous provisioner, then the reboot takes longer than this anyway.

Long and short, if we accept that we must now use "skip_clean" it makes 0.11.0 feel like a bit of a step backward rather than an improvement over 0.10.2 with regard to handling of reboots. Clearly, the need to handle reboots is an unfortunate reality when installing updates. It would be nice if we could preserve the 0.10.2 behaviour which handled all of this rather nicely.


** Perfectly if you use the reboot command suggested by sprd-bena in #3487 and a "pause_before" in the next provisioner. @jinesh-choksi #3487 is where I dug that command out from after I was having issues with the reboot command I was using previously!

**Comment by rickard-von-essen** [comment]
> Packer 0.10.2 handled these reboots perfectly

Actually it doesn't. Before 0.11.0 when ssh gets disconnected it was silently ignored. We fixed this bug in #3966.

> The example template and scripts _[...]_

Yes my suggestion require it to always assume a reboot. In general it's best to avoid reboots when creating a image and postpone booting a new kernel when a new instance is spawned from the image, if possible. 

> Long and short, if we accept that we must now use "skip_clean" it makes 0.11.0 feel like a bit of a step backward rather than an improvement over 0.10.2 with regard to handling of reboots. _[...]_

Unfortunately the handling of reboot was unreliable before 0.10.2 and errors could be silently ignore. When we fixed that it requires as stricter handling of reboots.

> While "skip_clean" may prove to be a valid workaround, it means we are left with Packer build scripts in the final image which is very undesirable.

Yes but this is most likely how it (silently) was working before, depending on timing. It's easy to clean it up in the next step. But usually it's best to end provisioning with a cleanup script that will wipe some dirs like `/tmp` and `/var/cache` etc.

@mwhooker Anything to add, do you see any improvements we can/should do regarding handling of reboots after #3966?

**Comment by DanHam** [comment]
OK. So from _my perspective_ 0.10.2 handled reboots perfectly - I was able to reboot, Packer would pick up where it left off on the other side and eventually make it all the way through to producing a usable image.

I'm sure that the fixes implemented in #3966 greatly improves Packer overall, but reboots in Packer versions -ge 0.11.0 now cause a build that worked previously to error and fail. With the greatest of respect, it could hardly be claimed that handling of reboots is now reliable post 0.10.2. 

It may be a small thing, but it's fairly typical for a new kernel to be installed when updating GNU/Linux based machines.  As such, post update and reboot I like to remove old kernels. Clearly, this would not be possible without a reboot. If we accept that reboots should be avoided, it places limits on the resulting images that we can create with Packer.

From my admittedly limited understanding, the issue we have found occurs post execution of the script. At this point Packer has determined that the communication channel it was using has been closed (by the reboot) and, instead of waiting for an exit code from the script, it proceeds to the next step of the process which is the 'clean-up' step. Could the logic that was used to determine the communication channel has gone down not be extended to the cleaning phase of the provisioner? In other words would it be possible to _wait_ on the return of a communication channel before attempting to perform any clean up tasks?

Again, many thanks for taking the time to look into this. It is much appreciated.

Dan

**Comment by rickard-von-essen** [comment]
After some additional thought on this maybe we should have a `posix-restart` provisioner, similar to [`windows-restart`](https://www.packer.io/docs/provisioners/windows-restart.html) that could handle this easily and gracefully.

**Comment by mwhooker** [comment]
yeah I agree a restart provisioner would make this more fool-proof, and may be the way to go.

This wasn't a failure i was anticipating when we made that change, though. Would definitely like to make this work by default, or at least offer a viable workaround. 

Trying to understand what exactly changed to introduce this behavior. I suppose, previously, it would silently disconnect and then try to reconnect over the reboot period. Perhaps now it doesn't try to reconnect once it's noticed that it disconnected. Does that sound correct? Maybe we could have it try to reconnect if `expect_disconnect` is `true`.

**Comment by rickard-von-essen** [comment]
@mwhooker Why does the provisioning fail when the clean up is disconnected? Should it use the same logic as the rest and silently ignore disconnects?

**Comment by mwhooker** [comment]
ah yeah, that's probably where the failure is. let me see if I can fix

**Comment by mwhooker** [comment]
the above pr should fix it.

as I mentioned in the PR, I don't think cleanup ever worked correctly if it happened immediately after a reboot. This should be Fixed Forever™️ now.

Thanks for the report @DanHam. Please let me know if you get a chance to test the PR. Will merge it in soon

**Comment by DanHam** [comment]
Fantastic!  @mwhooker I've just tested with the merged PR and everything is working fine.

One very welcome consequence of the fix is that it negates the need for 'pause_before' in the next provisioner. Since we are now held by the cleanup phase of the rebooting provisioner, Packer is effectively prevented from attempting to run the next provisioner and then being cut off in the middle of doing so by the pending reboot. Put another way, so long as there is no 'skip_clean' in the rebooting provisioner, there is now no need for the 'pause_before' in the next provisioner. 

Additionally, in the event that the provisioner _doesn't_ actually reboot the machine, everything just carries on as normal - and can now do so _without_ the need for any arbitrarily enforced pause. 

All in all, this fix has simplified reboots from an end user perspective which is great - reboots now just work!

> After some additional thought on this maybe we should have a posix-restart provisioner, similar to windows-restart that could handle this easily and gracefully.

Having a purpose built restart provisioner for *nix boxes at our disposal certainly couldn't hurt. However, if it's possible, I would really like the ability to 'talk to Packer' from my provisioning scripts. This doesn't have to be complicated - it could be as simple as creating a file in a specific, agreed upon location, that would then cause Packer to behave in a certain manner. For instance, rather than issuing a reboot command from my script, I could create ```/tmp/packer_reboot``` just before the script finished. Packer would spot this, wait for the script to exit, remove the file and then reboot the box. This would relieve Packer of the need to guess at what the (pesky!) end user was trying to do and has the possibility of simplifying the required logic within the provisioner considerably.

Thinking further, if we _could_ enable a way to talk back to Packer from our provisioning scripts it starts to open up some interesting possibilities. One such possibility would be a looping Windows restart provisioner - a '_Loopy Windows_ Restart' provisioner if you will. 

As you are no doubt aware, a major problem encountered when installing updates into a Windows box is that you never know in advance how many times you are going to need to reboot. In effect you go through the detect/download/install/reboot cycle until Windows Update tells you there are no more updates to apply. The current method most builders employ of applying Windows updates from a 'First Logon' script has some issues - Firstly, there is no visibility into the process from Packers perspective (since allowing access to WinRM is intentionally deferred until all updates are installed and all reboots are finished). Secondly, running the same script over and over across reboots requires its own, slightly awkward hack. 

This process could be greatly simplified if we could, for instance simply create a ```%windir%\Temp\Packer_Reboot_and_Repeat``` file just before the update script exited. Packer would then reboot the machine for us and instead of continuing on to the next provisioner, re-run the current one for us. Effectively this would give us the ability to perform the detect/download/install/reboot cycle, but hand off responsibilty for rebooting and re-running of the update script to the Packer provisioner. Clearly the logic required to detect, download and install updates and then detect and signal when a reboot is required would remain the responsibility of the scripts.

Just some thoughts... apologies if I've rambled on a bit there!

Again, many thanks for the fix and all your efforts.

Dan
